code/espurna/wifi.ino

```diff
@@ -687,31 +739,11 @@ void wifiLoop() {
         _wifiCheckAP();
     }
 
-    if (!wifiConnected()) {
-        return;
-    }
-
-    // Gratuitous arp
-    static unsigned long last_arp = 0;
-    if (millis() - last_arp < 5000) {
-        return;
-    }
-
-    unsigned long start = millis();
-    netif *n = netif_list;
-    bool res = false;
-    while (n) {
-        if ((n->flags & NETIF_FLAG_LINK_UP) || (n->flags & NETIF_FLAG_UP)) {
-            DEBUG_MSG_P(PSTR("sending arp...\n"));
-            etharp_gratuitous(n);
-            res = true;
-        } else {
-            DEBUG_MSG_P(PSTR("netif not yet up...\n"));
+    #if WIFI_GRATUITOUS_ARP_SUPPORT
+        // Only send out gra arp when in STA mode
+        if (_wifi_gratuitous_arp_interval && ((WiFi.getMode() & WIFI_AP) == 0)) {
+            _wifiSendGratuitousArp(_wifi_gratuitous_arp_interval);
```

Comment:
With STA+AP mode the WiFi radio is always on, which makes it less likely ARP requests may be missed.
But it may still be useful to send it if you get a (new) IP address.
Just make sure to only send it to the STATION_IF if you want to distinguish this.

N.B. the AP and STA use different MAC addresses, so it should be no problem to send out Gratuitous ARP to the AP interface. (as long as you're sending the right IP of course :) )

Reply:
This part is copied from the idf, basically trying to emulate GARP flag that they check. No flag here, so STATION_IF seems like a solution.